cpp/src/arrow/filesystem/s3fs_test.cc

```diff
@@ -590,6 +590,19 @@ class TestS3FS : public S3TestMixin {
     AssertObjectContents(client_.get(), "bucket", "somefile", "new data");
   }
 
+  void TestOpenOutputStreamCloseAsyncDestructor() {
+    std::shared_ptr<io::OutputStream> stream;
+    ASSERT_OK_AND_ASSIGN(stream, fs_->OpenOutputStream("bucket/somefile"));
+    ASSERT_OK(stream->Write("new data"));
+    // Destructor implicitly closes stream and completes the multipart upload.
+    // GH-37670: Testing it doesn't matter whether flush is triggered asynchronously
+    // after CloseAsync or synchronously after stream.reset().
```

Comment:
```suggestion
    // after CloseAsync or synchronously after stream.reset() since we're just
    // checking that `closeAsyncFut` keeps the stream alive until completion
    // rather than segfaulting on a dangling stream
```